openedx/core/djangoapps/credit/signature.py

```diff
@@ -27,18 +27,47 @@
 log = logging.getLogger(__name__)
 
 
+def _encode_secret(secret, provider_id):
+    """
+    Helper function for encoding text_type secrets into ascii.
+    """
+    try:
+        secret.encode('ascii')
+    except UnicodeEncodeError:
+        secret = None
+        log.error(u'Shared secret key for credit provider "%s" contains non-ASCII unicode.', provider_id)
+
+    return secret
+
 def get_shared_secret_key(provider_id):
     """
-    Retrieve the shared secret key for a particular credit provider.
+    Retrieve the shared secret for a particular credit provider.
+
+    It is possible for the secret to be stored in 2 ways:
+    1 - a key/value pair of provider_id and secret string
+        {'cool_school': '123abc'}
+    2 - a key/value pair of provider_id and secret dict (which should contain old
+        and new keys with their respective values)
+        {'cool_school':
+          {
+          'previous_key': '987zyx',
```

Comment:
I would use an array of valid keys instead of any names or explicitly expected keys because otherwise you will need to circle back and change the structure after you have completed a rotation as opposed to just removing the old key.

Reply:
For the signature checking, i've actually made the iteration over the dictionary "key agnostic" by way of dict.values()

The expectation is to send the new one, while we accept both back when 2 keys are present. If we have a list of keys (taken from the config), how should we make clear in the config which key is used for sending credit? I thought the explicit naming might be useful for that, so future SRE folks X time from now could say "ah okay, this is the key their currently using" without question.

The difference is `keys[0]` vs `keys['current_key']`

What are your thoughts?

Reply:
Oh, and sorry to be clear, when we remove the previous key, this code shouldn't need to change

Reply:
I can see you POV though... there's something nice about it just being a list of keys... we would just need to make sure the ordering is correct in the config, and also make a note here in the code what the 0th index of a credit provider's keys is referring to (new key, and not old key)

Reply:
Ah - ok I get what you're saying I forgot that we need to specifically use the old key for sending

Reply:
I was hoping to minimize the complexity of the rotation - but yeah having the order matter is error prone in a different way.

Reply:
Cool, LGTM now - I don't want to thumb it as I feel it would behoove us to get a review from someone on arch, but this comment is addressed - thanks Chris